### PR TITLE
Closes #12281 - a bug in cuda::pow with negative base values

### LINKED ISSUE #12281: `cv::cuda::pow` cannot handle negative values properly
##### System information (version)
- OpenCV => 4.0.0-pre
- Operating System / Platform => Windows 64 Bit (perhaps others)
- Compiler => Visual Studio 2013  (perhaps others)
- WITH_CUDA

##### Detailed description
Results of `cv::cuda::pow` with negative input and odd powers (e.g. 3) are all zeros (in case of int-type input) or positive value (in case of float-type input).

##### Steps to reproduce
```.py
npMat = np.arange(-2, 3) #[-2 -1  0  1  2]
cuMat = cv.cuda_GpuMat()
cuMat.upload(npMat)

print(npMat)
print(npMat**3)
print(cv.cuda.pow(cuMat, 3).download().ravel())
```
Note that I am using Python wrapper from #12234, but the bug should occur in C++ as well

##### Expected result
```
[-2 -1  0  1  2]
[-8 -1  0  1  8]
[-8 -1  0  1  8]
```

##### Actual result
```
[-2 -1  0  1  2]
[-8 -1  0  1  8]
[0 0 0 1 8] #for int64, int32, int16, int8 input types
#OR
[ 8.  1.  0.  1.  8.] #for float32, float64 input types
```

### PR DESCRIPTION
**Pull request by cv3d**
resolves #12281

```
force_builders=Custom
docker_image:Custom=ubuntu-cuda:16.04
```

### CODE REVIEW

File: modules/cudaarithm/src/cuda/math.cu [hunk 1]
```diff
@@ -278,43 +278,26 @@ namespace
 {
     template<typename T, bool Signed = numeric_limits<T>::is_signed> struct PowOp : unary_function<T, T>
     {
-        float power;
+        T power;
```
Comment (alalek) [inline modules/cudaarithm/src/cuda/math.cu]: `T` can be `uint8_t`
`power` can be `0.5f`.

"power" changes should be reverted.

**Comment by alalek** [comment]
As a "bugfix" this patch should [go](https://github.com/opencv/opencv/wiki/How_to_contribute#faq) into 3.4 branch. We will merge changes from 3.4 into master regularly (weekly/bi-weekly).

So, please:
- change "base" branch of this PR: master => 3.4 (use "Edit" button near PR title)
- rebase your commits from master onto 3.4 branch. For example:
    `git rebase -i --onto upstream/3.4 upstream/master`
    (check list of your commits, save and quit (Esc + "wq" + Enter)
    where `upstream` is configured by following [this GitHub guide](https://help.github.com/articles/configuring-a-remote-for-a-fork/) and fetched (`git fetch upstream`).
- push rebased commits into source branch of your fork (with `--force` option)

**Comment by cv3d** [comment]
@alalek I guess it is ready for a merge

File: modules/cudaarithm/src/cuda/math.cu [hunk 2]
```diff
@@ -282,39 +282,22 @@ namespace
...
-        __device__ __forceinline__ float operator()(float e) const
-        {
-            return __powf(::fabs(e), power);
-        }
-    };
-    template<> struct PowOp<double> : unary_function<double, double>
+    template<typename T> struct PowOp<T, false> : unary_function<T, T>
     {
-        double power;
+        float power;
```
Comment (alalek) [inline modules/cudaarithm/src/cuda/math.cu]: OpenCV should use "double" calculations (with double parameters) for double inputs.
It is OK to use float for integers and floats.
But for doubles we should use doubles anywhere (should not reduce them to floats).

Please restore "double" template specialization.

Comment (cv3d) [reply modules/cudaarithm/src/cuda/math.cu]: @alalek I made sure there is no precision issues in the last amendment.

**Pull request merged**